kcidb/tools/kcidb_match.py

```diff
@@ -0,0 +1,471 @@
+            if log_content and re.search(log_regex, log_content):
+                return True
+
+            return False
+
+        log_regex = pattern_object[key][0].get("log_regex")
+        if log_regex and key in kcidb_io_object and kcidb_io_object[key]:
+            log_excerpt = kcidb_io_object[key][0].get("log_excerpt", "")
+            log_url = kcidb_io_object[key][0].get("log_url", "")
+            return search_regex(log_regex, log_excerpt, log_url)
+        return False
+
+
+class IncidentGenerator:
+    """Class to generate incidents"""
+    def __init__(self, db_name=DB_NAME):
+        self.db = PatternDatabase(db_name)
+        self.validator = PatternValidator()
+
+    def create_incident(self, kcidb_io_object, issue_id, issue_version):
+        """Create and return an incident object"""
+        if 'tests' in kcidb_io_object and kcidb_io_object['tests']:
+            type_id_key = "test_id"
+            type_id_value = kcidb_io_object['tests'][0]['id']
+        elif 'builds' in kcidb_io_object and kcidb_io_object['builds']:
+            type_id_key = "build_id"
+            type_id_value = kcidb_io_object['builds'][0]['id']
+        else:
+            raise ValueError("The KCIDB IO object must contain at least "
+                             "one non-empty test or build")
+
+        unique_string = f"{issue_id}{issue_version}{type_id_value}"
+        incident_id = f"{ORIGIN}:{hashlib.sha256(unique_string.encode(
+                      )).hexdigest()}"
+
+        return {
+            'id': incident_id,
+            'origin': ORIGIN,
+            'issue_id': issue_id,
+            'issue_version': issue_version,
+            'present': True,
+            type_id_key: type_id_value,
+        }
+
+    def generate_incident_on_match(self, kcidb_io_object, issue_id,
+                                   issue_version, issue_pattern_object):
+        """Generate incident if issue pattern is found in a build/test
+        object"""
+        incidents = []
+
+        if self.validator.match_fields(issue_pattern_object, kcidb_io_object):
+            incident = self.create_incident(kcidb_io_object, issue_id,
+                                            issue_version)
+            incidents.append(incident)
+
+        return incidents
+
+    def generate_incidents_from_test(self, test):
+        """Generate incident from test object"""
+        kcidb_io_object = {"tests": [test._data],
+                           "builds": [test.build._data],
+                           "checkouts": [test.build.checkout._data]}
+        return self.generate_incidents_from_db(kcidb_io_object)
+
+    def generate_incidents_from_build(self, build):
+        """Generate incident from build object"""
+        kcidb_io_object = {"builds": [build._data],
+                           "checkouts": [build.checkout._data]}
+        return self.generate_incidents_from_db(kcidb_io_object)
+
+    def generate_incidents_from_db(self, kcidb_io_object):
```

Comment:
"""generate incidents by trying to match the kcidb_io_object against the patterns saved in DB_NAME"""